PublicGitArchive/pga-create/cmd/pga-create/discover.go

```diff
@@ -70,71 +70,78 @@ func reduceWatchers(stream io.Reader) map[int]int {
 		if err != nil {
 			fail(fmt.Sprintf("parsing watchers project ID \"%s\"", line[:commaPos]), err)
 		}
-		stars[projectID]++
+		stars[uint32(projectID)]++
 	}
 	return stars
 }
 
 type watchPair struct {
-	Value int
-	Key   int
+	Key, Value uint32
 }
 
-func writeWatchers(stars map[int]int, ids map[int][]int, path string) {
+func writeWatchers(stars map[uint32]uint32, ids map[uint32]bool, path string) {
 	pairs := make([]watchPair, 0, len(stars))
-	majors := map[int]bool{}
-	for key := range stars {
-		myids := ids[key]
-		// if it is a reference, go to the main repository
-		if myids[0] >= 0 {
-			key = myids[0]
-			myids = ids[key]
-		}
-		if majors[key] {
+	for k, v := range stars {
+		if !ids[k] {
 			continue
 		}
-		majors[key] = true
-		maxval := stars[key]
-		for _, id := range myids[1:] {
-			other := stars[id]
-			if other > maxval {
-				maxval = other
-			}
-		}
-		pairs = append(pairs, watchPair{Key: key, Value: maxval})
+		pairs = append(pairs, watchPair{Key: k, Value: v})
 	}
+	// Descending value order
 	sort.Slice(pairs, func(i, j int) bool {
-		return pairs[i].Value > pairs[j].Value // reverse order
+		return pairs[i].Value > pairs[j].Value
 	})
 	f, err := os.Create(path)
-	defer f.Close()
 	if err != nil {
 		fail("creating watchers file "+path, err)
 	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			fail("closing watchers file", err)
+		}
+	}()
+	gzf := gzip.NewWriter(f)
+	defer gzf.Close()
 	for _, pair := range pairs {
-		fmt.Fprintf(f, "%d %d\n", pair.Key, pair.Value)
+		_, err := fmt.Fprintf(gzf, "%d %d\n", pair.Key, pair.Value)
+		if err != nil {
+			fail("writing to watchers file", err)
+		}
 	}
 }
 
-func writeProjects(stream io.Reader, path string) map[int][]int {
+func writeProjects(stream io.Reader, path string) map[uint32]bool {
 	f, err := os.Create(path)
-	defer f.Close()
 	if err != nil {
 		fail("creating repositories file "+path, err)
 	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			fail("closing repositories file "+path, err)
+		}
+	}()
 	gzf := gzip.NewWriter(f)
 	defer gzf.Close()
 	scanner := bufio.NewScanner(stream)
 	skip := false
-	projects := map[string]int{}
-	ids := map[int][]int{}
+	ids := map[uint32]bool{}
 	for scanner.Scan() {
 		line := scanner.Text()
 		skipThis := skip
 		skip = line[len(line)-1:] == "\\"
 		if skipThis {
 			continue
 		}
+
+		// Skip deleted repositories
+		commaPos2 := strings.LastIndex(line, ",")
+		commaPos2 = strings.LastIndex(line[:commaPos2], ",")
+		commaPos1 := strings.LastIndex(line[:commaPos2], ",")
+		deletedFlag := line[commaPos1+1 : commaPos2]
+		if deletedFlag != "0" {
```

Comment:
How about the number of stars? If a repo is deleted, does it lose all the stars and the new one starts with 0?

Maybe we should take the maximum from the old and the new stars here.

Reply:
Yes, deleted repositories do not keep their stars if re-created. I think getting actual stars is the least surprising one, since it accurately reflects current status. This is particularly important for cases where the re-created repository is not the main one anymore, but a fork.